### DISCUSSION TITLE
babel runtime helpers added with .js extension causing ERR_PACKAGE_PATH_NOT_EXPORTED in subpath that fail when running node 16

### DISCUSSION THREAD

**Discussion by dagda1** [babel/babel#14951] [Q&A]
I am using `@babel/runtime` with the `useESModules: true` options and it is causing helpers to be emitted like this with the `.js` file extension

> `import "@babel/runtime/helpers/esm/extends.js";`

This causes the following error:

> [ERR_PACKAGE_PATH_NOT_EXPORTED]: Package subpath './helpers/esm/extends.js' is not defined by "exports" in `myproject/node_modules/@babel/runtime/package.json`

This is because this path conflicts with the `package.json` exports subpath pattern:

```json
"exports": {
  "./helpers/extends": [
  {
    "node": "./helpers/extends.js",
    "import": "./helpers/esm/extends.js",
    "default": "./helpers/extends.js"
   },
```

Removing the file extensions fixes this problem.  

Is there anyway to stop the file extensions getting added?

My workaround is this, which is extremely hacky, and I run after every build:

```sh
sed -i -e s/'extends.js'/'extends'/g ./dist/server/entry-server.js
sed -i -e s/'assertThisInitialized.js'/'assertThisInitialized'/g ./dist/server/entry-server.js
sed -i -e s/'inheritsLoose.js'/'inheritsLoose'/g ./dist/server/entry-server.js
sed -i -e s/'wrapNativeSuper.js'/'wrapNativeSuper'/g ./dist/server/entry-server.js
sed -i -e s/'taggedTemplateLiteralLoose.js'/'taggedTemplateLiteralLoose'/g ./dist/server/entry-server.js
```

**Comment by nicolo-ribaudo** [comment]
As a better workaround, if your environment supports `package.json#exports`, you can just disable `useESModules: true` and let Babel inject `import "@babel/runtime/extends"`.
- Node.js will load `@babel/runtime/helpers/extends.js`, which is the CJS version but in Node.js works exactly as the ESM version
- Non-node environments (bundlers or other runtimes) will skip the `"node"` key in `"exports"` and load `@babel/runtime/helpers/esm/extend.js`

**Reply by dagda1** [reply]
@nicolo-ribaudo thank you for the answer, but if I want to use ESM modules, so I set `useESModules: false` just about sums up javascript development at this time.

**Reply by nicolo-ribaudo** [reply]
Is there any reason why you need the ESM version of `@babel/runtime` instead of the CJS version? (I'm super in favor of ES modules, but I fail to understand what's the advantage in this specific case)

**Reply by dagda1** [reply]
More and more packages are going esm only so I've set all my package.json files to `"type": "module"`.  

Unfortunately a couple of things like `@vanilla-extract` use babel so I thought the `useESModule: true` option was the right path.

How wrong was I :).

**Reply by nicolo-ribaudo** [reply]
You can still compile your code to ESM when using Babel, but `useESModule: true` is just to select the type of `@babel/runtime`'s files :)

If you want to compile your own code to ESM rather than CJS,
- Make sure that you are not including `@babel/plugin-transform-modules-commonjs` in your config
- If you are using `@babel/preset-env`, enable its `modules: false` option (this is not necessary if you are also using a bundler)

**Comment by dagda1** [comment]
If `useESModule: true` is for the type of `@babel/runtime` helpers, then why is it creating import statements that  nukes its own code by not taking the `package.json` exports subpath into account?

i.e.

`import "@babel/runtime/helpers/esm/extends.js";`

nukes

```json
"exports": {
  "./helpers/extends": [
  {
    "node": "./helpers/extends.js",
    "import": "./helpers/esm/extends.js",
    "default": "./helpers/extends.js"
   },
```

It is a bug?

**Reply by nicolo-ribaudo** [reply]
It looks like a bug, and probably we never noticed it because now that we have `"exports"` that option is not really needed anymore (we might just remove it in Babel 8)

**Reply by dagda1** [reply]
@nicolo-ribaudo ok, thank you for taking the time to answer.  That has cleared it up, and I hope I did not come across rude.  I just hit so many hurdles with the switch to ESM and this was my final one

**Reply by dagda1** [reply]
I'm going to use your advice and use `useESModules: false` and put a link to this discussion above the code when I inevitably forget why I did it.

**Reply by nicolo-ribaudo** [reply]
I know that the migration to ESM can be painful, we have been slowing preparing to migrate our code and it's taking ages!

**Comment by dagda1** [comment]
@nicolo-ribaudo even with `useESModules: false`, it still seems to add the `.js` extension.  I'm reverting back to `sed` :).

**Reply by nicolo-ribaudo** [reply]
Can you share your full config, and the versions of the various packages?

**Comment by dagda1** [comment]
@nicolo-ribaudo I'm on the latest version of everything

```json
    "@babel/runtime": "7.19.0"
    "@babel/core": "^7.19.1",
    "@babel/plugin-proposal-nullish-coalescing-operator": "7.18.6",
    "@babel/plugin-proposal-optional-chaining": "7.18.9",
    "@babel/plugin-transform-regenerator": "7.18.6",
    "@babel/plugin-transform-runtime": "7.19.1",
    "@babel/preset-env": "7.19.1",
    "@babel/preset-react": "7.18.6",
    "@babel/runtime": "7.19.0",
```

I am using the rollup `@rollup/plugin-babel`

and my config looks like this:

```js
babel({
  babelHelpers: 'runtime',
  extensions: [...DEFAULT_EXTENSIONS, 'ts', 'tsx'],
  sourceType: 'module',
  babelrc: false,
  configFile: false,
  presets: [
    [
      '@babel/preset-env',
      {
        exclude: [
          'transform-typeof-symbol',
          '@babel/plugin-transform-regenerator'
        ],
        modules: 'auto',
        useBuiltIns: 'entry',
        corejs: 3
      }
    ],
    [ [Function (anonymous)], { runtime: 'automatic' } ]
  ],
  exclude: [ /node_modules\/react/, /\/node_modules\/(core-js)\// ],
  sourceType: 'module',
  plugins: [
    'babel-plugin-annotate-pure-calls',
    [
      '@babel/plugin-transform-runtime',
      {
        corejs: false,
        helpers: true,
        version: '7.19.0',
        regenerator: false,
        useESModules: false
      }
    ],
    '@babel/plugin-syntax-dynamic-import',
    '@loadable/babel-plugin',
    '@babel/plugin-proposal-optional-chaining',
    '@babel/plugin-proposal-nullish-coalescing-operator',
    '@vanilla-extract/babel-plugin'
  ]
 });
```

**Comment by JLHwung** [comment]
> import "@babel/runtime/helpers/esm/extends.js";

Does it come from node_modules? Maybe it is compiled by an older `runtime-transform`? Specifically, the `runtime-transform` 7.7.5 will add `.js` extension to the polyfill imports but we then reverted that behaviour in 7.7.6.

We have a similar test with `useESModules: true` and as you can see Babel does not add extensions:

https://github.com/babel/babel/blob/40dfe9ac463a231b3b0fc6a50365316debe15d06/packages/babel-plugin-transform-runtime/test/fixtures/dual-babel-runtime/useESModules-modern-runtime/output.mjs#L1-L6

Can you provide a reproduction repo?
